src/mongocxx/test/database.cpp

```diff
@@ -492,4 +492,95 @@ TEST_CASE("Database integration tests", "[database]") {
     }
 }
 
+// As C++11 lacks generic lambdas, and "ordinary" templates can't appear at block scope,
+// we'll have to define our helper for the serviceId tests here. This implementation would
+// be more straightforward in newer versions of C++ (C++14 and on allow generic lambdas),
+// see CXX-2350 (migration to more recent C++ standards); our C++17 optional<> implmentation
+// also has a few inconsistencies with the standard, which appear to vary across platforms:
+template <typename EventT>
+struct check_service_id {
+    const bool expect_service_id;
+
+    check_service_id(const bool expect_service_id) : expect_service_id(expect_service_id) {}
+
+    void operator()(const EventT& event) {
+        INFO("checking for service_id()")
+        CAPTURE(event.command_name(), expect_service_id);
+
+        auto service_id = event.service_id();
+
+        if (expect_service_id)
+            CHECK(service_id);
+        else
+            CHECK_FALSE(service_id);
+    }
+};
+
+TEST_CASE("serviceId presence depends on load-balancing mode") {
+    // Repeat this test case for a situation in which service_id should and should not be returned:
+    auto expect_service_id = GENERATE(true, false);
+
+    instance::current();
+
+    auto client_opts = test_util::add_test_server_api();
+
+    // Set Application Performance Monitoring options (APM):
+    mongocxx::options::apm apm_opts;
+    apm_opts.on_command_started(
+        check_service_id<mongocxx::events::command_started_event>{expect_service_id});
+    apm_opts.on_command_succeeded(
+        check_service_id<mongocxx::events::command_succeeded_event>{expect_service_id});
+    apm_opts.on_command_failed(
+        check_service_id<mongocxx::events::command_failed_event>{expect_service_id});
+
+    // Set up mocking for mongoc_apm_command_started_get_service_id:
+    auto apm_command_started_get_service_id =
+        libmongoc::apm_command_started_get_service_id.create_instance();
+    auto apm_command_succeeded_get_service_id =
+        libmongoc::apm_command_succeeded_get_service_id.create_instance();
+    auto apm_command_failed_get_service_id =
+        libmongoc::apm_command_failed_get_service_id.create_instance();
+
+    // Set up mocked functions that DO emit a service_id:
+    if (expect_service_id) {
+        // Return a bson_oid_t with data where the service_id has some value:
+        struct {
+            bson_oid_t* operator()(const void*) {
+                static bson_oid_t tmp = {0x65};
+                return &tmp;
+            }
+        } make_service_id_bson_oid_t;
+
+        // Add forever() so that the mock function also extends to endSession (thanks, kalbertson!):
```

Comment:
```suggestion
        // Add forever() so that the mock function also extends to all commands sent by the client, like "endSessions":
```

I suggest removing the thanks, since I can see it as being confusing to future readers. But you are very welcome :smile:. It is not obvious that "endSessions" is sent internally by clients, and that it appears in command monitoring events.

Reply:
Done-- but it will remain "morally there". ;-D